api/src/test/java/org/pmiops/workbench/user/DevUserRegistrationServiceTest.java

```diff
@@ -0,0 +1,84 @@
+
+import com.google.api.services.oauth2.model.Userinfoplus;
+import java.util.Optional;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.pmiops.workbench.db.dao.UserService;
+import org.pmiops.workbench.db.model.DbInstitution;
+import org.pmiops.workbench.db.model.DbUser;
+import org.pmiops.workbench.db.model.DbVerifiedInstitutionalAffiliation;
+import org.pmiops.workbench.exceptions.BadRequestException;
+import org.pmiops.workbench.google.DirectoryService;
+import org.pmiops.workbench.institution.InstitutionService;
+import org.pmiops.workbench.institution.VerifiedInstitutionalAffiliationMapperImpl;
+import org.pmiops.workbench.model.Institution;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.TestConfiguration;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.context.annotation.Import;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+public class DevUserRegistrationServiceTest {
+
+  @MockBean private DirectoryService directoryService;
+  @MockBean private InstitutionService institutionService;
+  @MockBean private UserService userService;
+
+  private Userinfoplus userInfo;
+  private ArgumentCaptor<DbVerifiedInstitutionalAffiliation> dbAffiliationCaptor =
+      ArgumentCaptor.forClass(DbVerifiedInstitutionalAffiliation.class);
+
+  @Autowired private DevUserRegistrationService service;
+
+  @TestConfiguration
+  @Import({DevUserRegistrationServiceImpl.class, VerifiedInstitutionalAffiliationMapperImpl.class})
+  static class Configuration {}
+
+  @Before
+  public void setUp() {
+    userInfo = new Userinfoplus().setEmail("gjordan@fake-research-aou.org");
+  }
+
+  @Test
+  public void testCreateUserFromUserInfo() {
+    // Tests the happy path: a contact email and matching institution are both found, allowing us
+    // to register the dev user.
+    DbUser dbUser = new DbUser();
+    dbUser.setContactEmail("gregory.jordan.123@gmail.com");
+    dbUser.setUsername("gjordan@fake-research-aou.org");
+    DbInstitution dbInstitution = new DbInstitution();
+    dbInstitution.setShortName("Google");
+
+    when(directoryService.getContactEmailFromGSuiteEmail(eq("gjordan@fake-research-aou.org")))
+        .thenReturn(Optional.of("gregory.jordan.123@gmail.com"));
+    when(institutionService.getFirstMatchingInstitution("gregory.jordan.123@gmail.com"))
+        .thenReturn(Optional.of(new Institution().shortName("Google")));
+    when(institutionService.getDbInstitutionOrThrow(eq("Google"))).thenReturn(dbInstitution);
+    when(userService.createUser(
+            eq(userInfo), eq("gregory.jordan.123@gmail.com"), dbAffiliationCaptor.capture()))
+        .thenReturn(dbUser);
+
+    service.createUserFromUserInfo(userInfo);
+
+    assertThat(dbAffiliationCaptor.getValue().getInstitution().getShortName()).isEqualTo("Google");
```

Comment:
Are there no other side effects to verify here?

Reply:
The Mockito stubs above pretty much cover the function under test line-by-line... I couldn't really think of anything else that wouldn't be super redundant.

Reply:
OK. I was going to  see if any other fields on the  created user could  be validated, but it's mocks all  the way down.